### PR TITLE
arm: tfm: samples: psa level 1 certified sample

### PR DESCRIPTION
**Pull request by microbuilder**
**WORK IN PROGRESS, SHOULD NOT BE MERGED!**

This PR adds some boilerplate code for a PSA Certified Level 1
sample application that can be used during the PSA certification
process.

It implements the following features:

- A config struct is defined (`cfg_data`) that contains app and user
  configuration data that will be persisted to secure storage via the
  appropriate PSA FF API calls.
- A `psa` shell command is added that can be extended with required
  features, such as the ability to delete the above config record and
  reset the device to factory defaults.
- Basic logging support to handle common TF-M error codes (`applog.c`),
  based on the Zephyr logging subsystem.
- Basic encryption/decryption support showing how the TF-M crypto
  service can be used to encrypt and decrypt data transfers.

This sample depends on #16450.

Signed-off-by: Kevin Townsend <kevin.townsend@linaro.org>

### CODE REVIEW

File: samples/tfm_integration/psa_level_1/README.rst [hunk 1]
```diff
@@ -0,0 +1,145 @@
+.. _tfm_psa_level_1:
+
+TF-M PSA Level 1
+################
+
+Overview
+********
+This TF-M integration example can be used with an Arm v8-M supported board. It
+implements the RTOS vendor requirements for a PSA Certified Level 1 product.
```
Comment (dbkinder) [inline samples/tfm_integration/psa_level_1/README.rst]: add the link you defined below:
```
implements the RTOS vendor requirements for a `PSA Certified Level 1`_ product.
```

File: samples/tfm_integration/psa_level_1/README.rst [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+are used for the secure processing environment, with Zephyr running in the
+non-secure processing environment.
+
+As part of the standard build process, the secure bootloader (BL2) is built, as
+well as the secure TF-M binary and non-secure Zephyr binary images. The two
+application images are then merged and signed using the public/private key pair
+stored in the secure bootloader, so that they will be accepted during the
+image validation process at startup.
+
+All TF-M dependencies will be automatically cloned in the /ext folder, and
```
Comment (dbkinder) [inline samples/tfm_integration/psa_level_1/README.rst]: change to:
```
in the ``/ext`` folder
```

File: samples/tfm_integration/psa_level_1/README.rst [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+
+.. note:: This example is based upon trusted-firmware-m commit 0d822496e809132e040c0a5cd556b2d5dad01ea6
+
+Building and Running
+********************
+
+This project outputs startup status and info to the console. It can be built and
+executed on an MPS2+ configured for AN521 (dual-core ARM Cortex M33).
+
+At present, this sample is only compatible with Linux and OS X based systems,
```
Comment (dbkinder) [inline samples/tfm_integration/psa_level_1/README.rst]: Change to:
```
This sample will only build on a Linux or macOS development system (not Windows),
```

File: samples/tfm_integration/psa_level_1/README.rst [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+and has been tested on the following setups:
+
+- OS X 10.14.5 natively with gcc-arm-none-eabi-7-2018-q2-update
```
Comment (dbkinder) [inline samples/tfm_integration/psa_level_1/README.rst]: change ``OS X`` to ``macOS Mojave``

File: samples/tfm_integration/psa_level_1/README.rst [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+- Ubuntu 18.04 using Zephyr SDK 0.10.1
+
+On MPS2+ AN521:
+===============
+
+1. Build Zephyr with a non-secure configuration (``-DBOARD=mps2_an521_nonsecure``). This will also build TF-M in the ``ext/tfm`` folder if necessary.
```
Comment (dbkinder) [inline samples/tfm_integration/psa_level_1/README.rst]: Keep line length < 80 when possible

File: samples/tfm_integration/psa_level_1/README.rst [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+
+Using ``west``
```
Comment (dbkinder) [inline samples/tfm_integration/psa_level_1/README.rst]: All these lines (46-61) should be indented an additional three spaces to be part of step 1.

File: samples/tfm_integration/psa_level_1/README.rst [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+
+   cd <ZEPHYR_ROOT>/samples/tfm_integration/psa_level_1/
+   mkdir build
+   cd build
+   cmake -GNinja -DBOARD=mps2_an521_nonsecure ..
+   ninja -v
+
+2. Merge the TF-M secure and Zephyr non-secure images, and sign the final image
+
+.. code-block:: bash
```
Comment (dbkinder) [inline samples/tfm_integration/psa_level_1/README.rst]: same here.  Indent the code block and contents (line 65-67) by an additional three spaces.

File: samples/tfm_integration/psa_level_1/README.rst [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+
+   ./merge.sh
+
+3. Copy application binary files (mcuboot.bin and tfm_sign.bin) to ``<MPS2 device name>/SOFTWARE/``.
+4. Open ``<MPS2 device name>/MB/HBI0263C/AN521/images.txt``.
```
Comment (dbkinder) [inline samples/tfm_integration/psa_level_1/README.rst]: Let's merge some of these steps and fix some formatting issues:
```
4. Edit (e.g., with vim) the ``<MPS2 device name>/MB/HBI0263C/AN521/images.txt``
   file, and update it as shown below:

   .. code-block:: bash

      TITLE: Versatile Express Images Configuration File

      [IMAGES]
      TOTALIMAGES: 2 ;Number of Images (Max: 32)

      IMAGE0ADDRESS: 0x10000000
      IMAGE0FILE: \SOFTWARE\mcuboot.bin  ; BL2 bootloader

      IMAGE1ADDRESS: 0x10080000
      IMAGE1FILE: \SOFTWARE\tfm_sign.bin ; TF-M with application binary blob

5. Save the file, exit the editor, and reset the MPS2+ board.
```

File: samples/tfm_integration/psa_level_1/README.rst [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+
+Signing Images
+==============
+
+TF-M uses a secure bootloader (BL2) and firmware images must be signed
+with a private key before execution can be handed off by the bootloader. The
+firmware image is validated by the bootloader at startup using the public key,
+which is built into the secure bootloader.
+
+By default, tfm/bl2/ext/mcuboot/root-rsa-3072.pem is used to sign images.
```
Comment (dbkinder) [inline samples/tfm_integration/psa_level_1/README.rst]: It renders better if you enclose filenames with double back-ticks:
```
By default, ``tfm/bl2/ext/mcuboot/root-rsa-3072.pem`` is used to sign images.
The ``merge.sh`` script signs the TF-M + Zephyr binary using the ``.pem`` private key,
by calling ``imgtool.py`` to perform the actual signing operation.
```

File: samples/tfm_integration/psa_level_1/README.rst [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+merge.sh signs the TF-M + Zephyr binary using the .pem private key,
+calling imgtool.py` to perform the actual signing operation.
+
+In order to satisfy PSA Level 1 certification requirements, **You MUST replace
```
Comment (dbkinder) [inline samples/tfm_integration/psa_level_1/README.rst]: Change ``In order to`` to simply ``To``

File: samples/tfm_integration/psa_level_1/README.rst [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+   [INF] Starting bootloader
+   [INF] Swap type: none
+   [INF] Bootloader chainload address offset: 0x80000
+   [INF] Jumping to the first image slot
+   [Sec Thread] Secure image initializing!
+   ***** Booting Zephyr OS zephyr-v1.14.0-1616-g3013993758d4 *****
+   ...
+
+
+.. _TF-M build instruction:
```
Comment (dbkinder) [inline samples/tfm_integration/psa_level_1/README.rst]: You've defined these two external links, but don't use them anywhere in the document so they won't show up in our generated HTML.  Note what I did on line 9 above to add a link in the documentation to the document you identified.  We need to create similar references in the text to these two documents.

**Review by dbkinder — CHANGES_REQUESTED** [review]

File: boards/arm/mps2_an521/mps2_an521_nonsecure.dts [hunk 12]
```diff
@@ -45,8 +45,8 @@
 		reg = <0x28100000 0x100000>;
 	};
 
-	flash0: flash@100000 {
-		reg = <0x100000 0xDF00000>;
+	flash0: flash@100400 {
```
Comment (karl-zh) [inline boards/arm/mps2_an521/mps2_an521_nonsecure.dts]: This commit is from #16450 at [here](https://github.com/zephyrproject-rtos/zephyr/pull/16450/commits/f9d10e31e2436c2dc604f2a3fe43d8dd2b67e961) to resolve secure fault while jump into Zephyr.

**Review by karl-zh — COMMENTED** [review]
The Crypto test cases are leveraging the same non secure test cases from TFM. These test cases are designed for TFM crypto service. @microbuilder Is it fine to take advantage these integration test cases here?

**Comment by microbuilder** [comment]
Superseded by #19985, which takes into account the changes and feedback made in this PR.

This PR is being closed in favour of the newer version which integrates a shared TF-M build system, and will simplify the overall TF-M sample app review process.

**Pull request closed without merging**